torch/csrc/autograd/python_variable.cpp

```diff
@@ -413,6 +427,189 @@ static PyObject * THPVariable_device(THPVariable* self) {
+    ptype = THPVariable_result_ptype((PyObject*)self, in_args);
+    Py_INCREF(ptype);
+    base = THPVariableClass;
+    Py_INCREF(base);
+    in_no = PyTuple_GET_SIZE(in_args);
+    if (in_args == NULL) {
+        return NULL;
+    }
+    for (int i = 0; i < in_no; ++i){
+        obj = PyTuple_GET_ITEM(in_args, i);
+        if (PyObject_IsInstance(obj, base)){
+            auto& var = ((THPVariable*)obj)->cdata;
+            result_requires_grad |= var.requires_grad();
+            var.detach_();
+            new_obj = torch::utils::tensor_to_numpy(var.data());
+            Py_DECREF(obj);
+            PyTuple_SET_ITEM(in_args, i, new_obj);
+        }
+    }
+
+    // Down-cast Output Arguments to np.ndarray
+    out_no = 0;
+    key = PyUnicode_FromString("out");
+    if (kwargs != NULL && PyDict_Contains(kwargs, key) == 1){
+        out_args = PyDict_GetItem(kwargs, key);
+        if (PyTuple_Check(out_args) == 0){
+            out_args = PyTuple_Pack(1, out_args);
+        }
+        out_no = PyTuple_GET_SIZE(out_args);
+        for (int i = 0; i < out_no; ++i){
+            obj = PyTuple_GET_ITEM(out_args, i);
+            if (PyObject_IsInstance(obj, base)){
+                auto& var = ((THPVariable*)obj)->cdata;
+                var.detach_();
+                new_obj = torch::utils::tensor_to_numpy(var.data());
+                Py_DECREF(obj);
+                PyTuple_SET_ITEM(out_args, i, new_obj);
+            }
+        }
+        PyDict_SetItem(kwargs, key, out_args);
+    }
+    else{
+        out_args = PyTuple_New(((PyUFuncObject*)ufunc)->nout);
+        out_no = PyTuple_GET_SIZE(out_args);
+    }
+    Py_DECREF(key);
+
+    // Call ufunc with np.ndarray types
+    ufunc_method = PyObject_GetAttr(ufunc, method_name);
+    if (ufunc_method == NULL)
+        goto cleanup;
+    result = PyObject_Call(ufunc_method, in_args, kwargs);
+    Py_DECREF(ufunc_method);
+
+    // Result Error Checking
+    if (PyUnicode_CompareWithASCIIString(method_name, "at") == 0){
+        goto cleanup;
+    }
+    if (result == NULL || result == Py_NotImplemented)
+        goto cleanup;
+
+    // Up-Cast Results to ptype
+    if (((PyUFuncObject*)ufunc)->nout == 1){
+        if (PyTypeNum_ISBOOL(PyArray_DESCR((PyArrayObject*)result)->type_num)){
+            //Workaround, torch has no built-in bool tensor
+            new_obj = PyArray_CastToType((PyArrayObject*)result, PyArray_DescrFromType(NPY_UINT8), /*is_fortran=*/false);
+            Py_DECREF(result);
+            result = new_obj;
+        }
+        results = PyTuple_Pack(1, result);
+    }
+    else{
+        results = result;
+        Py_INCREF(results);
+    }
+    Py_DECREF(result);
+    for (int i = 0; i < out_no; ++i){
+        obj = PyTuple_GET_ITEM(out_args, i);
+        if(obj == NULL){
+            obj = PyTuple_GET_ITEM(results, i);
+        }
+        auto var = torch::utils::tensor_from_numpy(obj);
+        result_requires_grad &= at::is_floating_point(var);
```

Comment:
This looks... questionable.

Reply:
> "This looks... questionable."

I totally agree, I have fixed this so that requires_grad=False when calling a numpy `__array_ufunc__` because autograd is not possible.
This was a bogus workaround to pass test_multiplication_numpy_scalar in test_torch.py which tests.

Reply:
This was a bogus workaround to pass test_multiplication_numpy_scalar in test_torch.py